// smooch-core/lib/api/appUsersStripe.js
'use strict';

exports.__esModule = true;
exports.AppUsersStripeApi = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _base = require('./base');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @class AppUsersStripeApi
 * @extends BaseApi
 */

var AppUsersStripeApi = exports.AppUsersStripeApi = function (_BaseApi) {
    (0, _inherits3.default)(AppUsersStripeApi, _BaseApi);

    function AppUsersStripeApi() {
        (0, _classCallCheck3.default)(this, AppUsersStripeApi);
        return (0, _possibleConstructorReturn3.default)(this, _BaseApi.apply(this, arguments));
    }

    AppUsersStripeApi.prototype.updateCustomer = function updateCustomer(userId, token) {
        if (!token) {
            return _promise2.default.reject(new Error('Must provide a Stripe token.'));
        }

        var url = this.getFullURL('appUsers', userId, 'stripe', 'customer');
        return this.request('POST', url, {
            token: token
        }, {
            allowedAuth: ['jwt']
        });
    };

    AppUsersStripeApi.prototype.createTransaction = function createTransaction(userId, actionId, token) {
        if (!actionId) {
            return _promise2.default.reject(new Error('Must provide an action id.'));
        }

        var url = this.getFullURL('appUsers', userId, 'stripe', 'transaction');

        var body = {
            actionId: actionId
        };

        if (token) {
            (0, _assign2.default)(body, {
                token: token
            });
        }

        return this.request('POST', url, body);
    };

    return AppUsersStripeApi;
}(_base.BaseApi);
